### PR TITLE
Window frame support

### PR DESCRIPTION
**Pull request by max-hoffman**
This PR adds sql.WindowFrame and sql.WindowFramers to represent
window frames from AST -> plan -> exec.

Every unique and valid combination of window frame
Start and End bounds has a plan node (sql.WindowFrame)
and a framer (sql.WindowFramer) pair. The frames, framers,
and constructor functions that map from AST -> plan and
plan -> exec are generated by optgen.

### CODE REVIEW

File: optgen/README.md [hunk 1]
```diff
@@ -21,3 +21,17 @@ when the opportunities arise.
 If we end up using more of cockroach DB's optimizer, they codegen their
 expressions, normalization rules, exploration rules, and execution code
 stem from this general setup.
+
+## Generators
```
Comment (max-hoffman) [inline optgen/README.md:25]: @zachmu summary of codegen

File: sql/core.go [hunk 2]
```diff
@@ -166,6 +166,19 @@ type WindowFramer interface {
 	//SlidingInterval(ctx Context) (WindowInterval, WindowInterval, WindowInterval)
 }
```
Comment (max-hoffman) [inline sql/core.go:168]: @zachmu this is important, will add doc comments

File: sql/expression/function/aggregation/window_framer.go [hunk 3]
```diff
@@ -255,3 +257,162 @@ func (f *GroupByFramer) Interval() (sql.WindowInterval, error) {
 func (f *GroupByFramer) SlidingInterval(ctx sql.Context) (sql.WindowInterval, sql.WindowInterval, sql.WindowInterval) {
 	panic("implement me")
 }
+
```
Comment (max-hoffman) [inline sql/expression/function/aggregation/window_framer.go:260]: @zachmu this is the point of failure for ROWS frame correctness

File: sql/parse/window_frame_factory.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
+package parse
+
+import (
+	ast "github.com/dolthub/vitess/go/vt/sqlparser"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+//go:generate optgen -out window_frame_factory.og.go -pkg parse frameFactory window_frame_factory.go
+
```
Comment (max-hoffman) [inline sql/parse/window_frame_factory.go:10]: @zachmu this is the bottleneck for AST -> plan node correctness

File: sql/expression/function/aggregation/window_framer.og.go [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+		rangeFramerBase{
+			startFollowingExpr: startFollowingExpr,
+			unboundedFollowing: unboundedFollowing,
+		},
+	}, nil
+}
+
+func NewFramer(frame sql.WindowFrame) (sql.WindowFramer, error) {
+	switch frame.Op() {
+	case 0:
```
Comment (zachmu) [inline sql/expression/function/aggregation/window_framer.og.go:607]: Better to generate each of these framers their own NewFramer method that you call

File: sql/plan/window_frame.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import (
+	"fmt"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+//go:generate optgen -out window_frame.og.go -pkg plan frame window_frame.go
+
+// windowFrameBase maintains window frame bounds if
+// set for an Window's over clause.
```
Comment (zachmu) [inline sql/plan/window_frame.go:12]: a Window's

**Review by zachmu — APPROVED** [review]
Seems like a sound approach, just one major suggestion.

This needs more tests, probably generated as well.

File: sql/expression/common.go [hunk 7]
```diff
@@ -115,3 +119,33 @@ func Dispose(e sql.Expression) {
 		return true
 	})
 }
+
+// ToOffset extracts a non-negative integer from an expression.Literal, or errors
+func ToOffset(e sql.Expression) (int, error) {
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/expression/common.go:124]: yeah that's a lot nicer

File: sql/expression/function/aggregation/window_framer.go [hunk 8]
```diff
@@ -255,3 +257,161 @@ func (f *GroupByFramer) Interval() (sql.WindowInterval, error) {
...
+	partitionSet   bool
+
+	startOffset int
+	endOffset   int
+
+	unboundedFollowing bool
+	unboundedPreceding bool
+	startCurrentRow    bool
+	endCurrentRow      bool
+	startPrecedingExpr int
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/expression/function/aggregation/window_framer.go:276]: what if i called them both `startNPreceding`?

**Pull request merged**